### PR TITLE
SA recovery

### PR DESCRIPTION
**Pull request by ilovezfs**
Quoting @lundman from the commit message: "If zget attempts to create a new vnode with an invalid vtype (VNON) we attempt to lookup the correct value from the dirent zap. The SA is still incorrect but can be deleted. Code is enabled with zfs_recover set."

### CODE REVIEW

File: module/zfs/zfs_znode.c [hunk 1]
```diff
@@ -822,6 +822,49 @@ zfs_znode_alloc(zfsvfs_t *zfsvfs, dmu_buf_t *db, int blksz,
...
+			printf("ZFS: WARNING! objid %llu has invalid SA data, please restore from backup. (mode %x)\n",
+				   zp->z_id, (int)zp->z_mode);
+
+			uint64_t parent = zfsvfs->z_recover_parent;
+			if (parent)	{
+
+				zap_cursor_t zc;
+				zap_attribute_t *za;
+				int err;
+				uint64_t mask =  ZFS_DIRENT_OBJ(-1ULL);
```
Comment (ilovezfs) [inline module/zfs/zfs_znode.c]: What is the significance of -1 here?

File: module/zfs/zfs_znode.c [hunk 2]
```diff
@@ -822,6 +822,49 @@ zfs_znode_alloc(zfsvfs_t *zfsvfs, dmu_buf_t *db, int blksz,
...
+
+				za = kmem_alloc(sizeof (zap_attribute_t), KM_SLEEP);
+				for (zap_cursor_init(&zc, zfsvfs->z_os, parent);
+					 (err = zap_cursor_retrieve(&zc, za)) == 0;
+					 zap_cursor_advance(&zc)) {
+					if ((za->za_first_integer & mask) == (zp->z_id & mask)) {
```
Comment (ilovezfs) [inline module/zfs/zfs_znode.c]: What is this testing?

**Comment by ilovezfs** [comment]
Are there any soft panics that exist (or that we could add) corresponding to this fix, where we could add a suggestion to set zfs_recover and/or add a url pointing to info on the error?

File: module/zfs/zfs_znode.c (hunk 2 above)
Comment (lundman) [inline module/zfs/zfs_znode.c]: It is the mask passed along for comparison, bit AND against the id. As for why you would want to mask bits in the objectID, I wouldn't have a clue. It is already part of zap_search_value() from whence the code came, and I merely set it to the same value the sources always use.

File: module/zfs/zfs_znode.c (hunk 1 above)
Comment (lundman) [inline module/zfs/zfs_znode.c]: It is the mask passed along for comparison, bit AND against the id. As for why you would want to mask bits in the objectID, I wouldn't have a clue. It is already part of zap_search_value() from whence the code came, and I merely set it to the same value the sources always use.

**Comment by lundman** [comment]
No soft panics that I could trigger. It was quite happy to finally let me unlink everything bad. But it left us with a concern that with zdb, we could dump the pool, which would show debris left behind with objID of that of the bad symlink. It is unknown whether this is a problem or not.

**Pull request merged**